source/base/ThreadPool.ooc

```diff
@@ -1,111 +1,203 @@
 use ooc-collections
+use ooc-base
 
 import structs/LinkedList
 import threading/Thread
+import threading/native/ConditionUnix
+import os/Time
 
-ThreadJob: class {
-	_body: Func
-	init: func (=_body)
-	_execute: func { this _body() }
-	run: virtual func {
-		this _execute()
-		this free()
-	}
+_TaskState: enum {
+	Unfinished
+	Finished
+	Cancelled
 }
 
-SynchronizedThreadJob: class extends ThreadJob {
-	_finishedCondition := WaitCondition new()
-	_mutex := Mutex new()
-	_finished := false
-	finished ::= this _finished
-	_freeOnCompletion := false
-	init: func (body: Func) { super(body) }
+_Task: class {
+	_mutex: Mutex
+	_action: Func
+	_state : _TaskState
+	_freeDirectly: Bool
+	init: func (=_action, =_mutex) {
+		_state = _TaskState Unfinished
+		_mutex = Mutex new()
+	}
 	free: override func {
-		this _mutex lock()
-		if (!this _finished) {
-			this _freeOnCompletion = true
-			this _mutex unlock()
-		} else {
-			this _mutex unlock()
-			this _mutex destroy()
-			this _finishedCondition free()
+		if (this _state != _TaskState Unfinished)
+			this _freeDirectly = true
+		else
 			super()
-		}
 	}
-	run: override func {
-		this _execute()
-		this _mutex lock()
-		this _finished = true
-		if (this _freeOnCompletion) {
-			this _mutex unlock()
+	run: virtual func (=_mutex) {
+		this _action()
+		_mutex lock()
+		if (this _state != _TaskState Cancelled)
+			this _state = _TaskState Finished
+		_mutex unlock()
+		if (this _freeDirectly)
 			this free()
-		} else {
-			this _mutex unlock()
-			this _finishedCondition broadcast()
-		}
 	}
-	wait: func {
+}
+
+_ResultTask: class <T> extends _Task {
+	_result: Cell<T>
+	_resultAction: Func -> T
+	init: func (=_resultAction, =_mutex) {
+		this _state = _TaskState Unfinished
+	}
+	free: func {
+		if (this _state != _TaskState Unfinished)
+			this _freeDirectly = true
+		else
+			super()
+	}
+	run: override func (=_mutex) {
+		temporary := Cell<T> new(this _resultAction())
 		this _mutex lock()
-		if (!this _finished)
-			this _finishedCondition wait(this _mutex)
+		this _result = temporary
+		if (this _state != _TaskState Cancelled)
+			this _state = _TaskState Finished
+		if (this _freeDirectly)
+			this free()
 		this _mutex unlock()
 	}
 }
 
+_TaskPromise: class extends Promise {
+	_task: _Task
+	init: func (=_task)
+	free: override func {
+		this _task free()
+		super()
+	}
+	wait: override func -> Bool {
+		status := false
+		while (true) {
+			this _task _mutex lock()
+			if (this _task _state != _TaskState Unfinished) {
+				status = (this _task _state == _TaskState Finished)
+				this _task _mutex unlock()
+				break
+			}
+			this _task _mutex unlock()
+			Time sleepMilli(10)
+		}
+		status
+	}
+	cancel: override func -> Bool {
```

Comment:
Ideally, this would interrupt/cancel the thread and have it move on to the next task in the queue (if any). Haven't yet figured out a nice way to do this though.